### PR TITLE
refactor: create shared client registry for sharing

### PR DESCRIPTION
**Pull request by aaron-steinfeld**
## Description
The idea here is to create an extendable shared client registry so clients can be shared across enrichers,
 rather than having each enricher hold client configs independently and create their own. Only entity service previously had a sharing mechanism, and it wasn't testable or extendable, and had the strange behavior where we cached the client based on the first config that was provided, even though we required each enricher to provide its own config.

### Testing
Updated unit tests, built and verified functionality e2e

### CODE REVIEW

**Comment by tim-mwangi** [comment]
Seems like a great idea to me :)

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java
Comment (aaron-steinfeld) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java:24]: Didn't seem worth testing this class as it's just client building in the constructor

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/BackendEntityEnricherTest.java
Comment (aaron-steinfeld) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/BackendEntityEnricherTest.java:61]: Used real entitycache in the tests as they were already written to expect that (but implicitly, since it was previously retrieved internally via a static method. These tests could be improved by mocking that dependency, too.

File: hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java
Comment (aaron-steinfeld) [inline hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java:46]: I'm not sure what the parallelism expectations are on this class. It reads like we expect init to be called from multiple threads, but we only want a single `EnrichmentProcessor` instance. I'm not sure there are statics being used, do we expect there to be multiple `StructuredTraceEnrichProcessor` instances sharing one `EnrichmentProcessor`?

The init side of this will handle that fine, but the cleanup side - which prior to this PR was ignored and not doing any cleanup - would be problematic. We'd have to use something like a countdown latch to ensure we don't clean up until all instances are out of use.

Comment (aaron-steinfeld) [reply hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java:46]: Dug way back in the archives - looks like this was written back before this was written on kafka streams. So does someone with more context on kafka streams know if it's still needed?

**Comment by aaron-steinfeld** [comment]
Snyk issue resolved by https://github.com/hypertrace/service-framework/pull/25 (will merge a dependency update in here once that PR is merged)

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java
Comment (tim-mwangi) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java:24]: But we love big coverage numbers ;)

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/DefaultServiceEntityEnricherTest.java
Comment (tim-mwangi) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/DefaultServiceEntityEnricherTest.java:221]: Interesting

**Review by tim-mwangi — APPROVED** [review]

File: hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java
Comment (tim-mwangi) [reply hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java:46]: @surajpuvvada @buchi-busireddy

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/DefaultServiceEntityEnricherTest.java
Comment (aaron-steinfeld) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/DefaultServiceEntityEnricherTest.java:221]: Switching to use the mockito extension here (which uses strict stubs by default) highlighted a number of issues with the tests.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java
Comment (aaron-steinfeld) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/clients/DefaultClientRegistry.java:24]: Yeah, yeah me too. Not only are these statics a big pain to test, but they don't provide value since we've got no logic in here - just invocations to duplicate in the test.

**Review by skjindal93 — APPROVED** [review]

**Review by tim-mwangi — APPROVED** [review]

File: hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java
Comment (surajpuvvada) [reply hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java:46]: Yes there can be multiple instances. Typically you can imagine one per stream thread. I think the EnrichmentProcessor might be using a cache somewhere and my guess is it is to prevent duplicate caches across instances. I vaguely recall running into memory issues..

Comment (aaron-steinfeld) [reply hypertrace-trace-enricher/hypertrace-trace-enricher/src/main/java/org/hypertrace/traceenricher/trace/enricher/StructuredTraceEnrichProcessor.java:46]: So there are multiple instances of `StructuredTraceEnrichProcessor` all sharing the same underlying instance of `EnrichmentProcessor`? What could go wrong 🙄 !

Given that, I'll remove the cleanup happening now (which is bad, but mirrors the previous implementation). I think separately, it's worth allowing the creation of multiple instances here, since the client registry allows sharing caches across instances now.

**Review by tim-mwangi — APPROVED** [review]

**Pull request merged**
